### PR TITLE
Optimize copy!(::Dict, ::Dict)

### PR DESCRIPTION
**Pull request by tkf**
This PR implements two optimizations for `copy!(dst::Dict, src::Dict)`. This came up in another PR for implementing `setindex` for `Dict`: https://github.com/JuliaLang/julia/pull/33495#discussion_r357008424

### Case 1: `copy!(dst::D, src::D) where {D <: Dict}`

This is a straightforward case where you can copy properties as-is:

```julia
function unsafe_copy!(dst::D, src::D) where {D <: Dict}
    copy!(dst.vals, src.vals)
    copy!(dst.keys, src.keys)
    copy!(dst.slots, src.slots)

    dst.ndel = src.ndel
    dst.count = src.count
    dst.age = src.age
    dst.idxfloor = src.idxfloor
    dst.maxprobe = src.maxprobe
    return dst
end


using BenchmarkTools
d = Dict{Int,Int}(zip(1:10^5, 1:10^5))
c = Dict{Int,Int}()
@btime copy!($c, $d)
@btime unsafe_copy!($c, $d)
```

prints

```
  2.146 ms (0 allocations: 0 bytes)
  299.572 μs (0 allocations: 0 bytes)
```

in Julia 1.3.

### Case 2: `copy!(dst::Dict{Kd}, src::Dict{Ks}) where {Ks,Kd>:Ks}`

This is the case you don't need to re-compute hashes:

```julia
using Base: isslotfilled

function unsafe_copy!(dst::Dict, src::Dict)
    resize!(dst.vals, length(src.vals))
    resize!(dst.keys, length(src.keys))

    svals = src.vals
    skeys = src.keys
    dvals = dst.vals
    dkeys = dst.keys
    @inbounds for i = src.idxfloor:lastindex(svals)
        if isslotfilled(src, i)
            dvals[i] = svals[i]
            dkeys[i] = skeys[i]
        end
    end

    copy!(dst.slots, src.slots)

    dst.ndel = src.ndel
    dst.count = src.count
    dst.age = src.age
    dst.idxfloor = src.idxfloor
    dst.maxprobe = src.maxprobe
    return dst
end


using BenchmarkTools
d = Dict{Int,Int}(zip(1:10^5, 1:10^5))
c = Dict{Any,Float64}()
@btime copy!($c, $d)
@btime unsafe_copy!($c, $d)
```

This prints

```
  4.030 ms (130549 allocations: 1.99 MiB)
  1.889 ms (99489 allocations: 1.52 MiB)
```

in Julia 1.3.

### CODE REVIEW

**Comment by thofma** [comment]
Does the case 2) assume that `hash(convert(T, x)) == hash(x)` for `x::S` and `S <: T`?

**Comment by rfourquet** [comment]
> Does the case 2) assume that hash(convert(T, x)) == hash(x) for x::S and S <: T?

Do you have an example where `x !== convert(T, x)` ? Unless `S == T`, `T` would be abstract, and I would expect `convert(T, x)` to just return `x` because `x` is already a `T`.

**Comment by thofma** [comment]
Why would `T` be abstract? I can define `convert` for any type I want  and the documentation does not mention that it has to preserve hashes.

So to answer your question:
```julia
julia> struct A
       end

julia> struct B
       end

julia> Base.convert(::A, ::Type{B}) = B()

julia> a = A(); convert(a, B) !== A
true
```
There are also methods of the form `convert(::S, ::Type{T})` for concrete `S, T` with `S !== T` in Base itself.

**Comment by rfourquet** [comment]
> Why would T be abstract?

When `S <: T` and `S != T`, generally `T` is abstract (there are probably some corner cases, I believe there is an issue where two types are not equal but represent the same set of values). If `S` is concrete and `S != T`, I think `T` is always abstract, but I'm not certain).

> There are also methods of the form convert(::S, ::Type{T}) for concrete S, T with S !== T in Base itself.

Besides the reversed order of the argument, yes there is no problem with that. My remark was based on the `S <: T` relation that you mentioned.

**Comment by rfourquet** [comment]
> I can define convert for any type I want and the documentation does not mention that it has to preserve hashes.

Correct, for example `x = rand(); hash(x) != hash(convert(Float16, x))` (most of the time).

**Comment by thofma** [comment]
Sorry, you are right.

So in the signature of the second case, either `Kd === Ks` or `Kd` is abstract and thus `convert(::Type{Kd}, x::Ks) === x`?

**Comment by rfourquet** [comment]
> So in the signature of the second case, either Kd === Ks or Kd is abstract and thus convert(::Type{Kd}, x::Ks) === x?

Yes. Even without considering the abstract/non-abstract distinction, if `Kd>:Ks`, then it is expected that `convert(Kd, x::Ks) === x` because `x` is already of the requested type (`x isa Ks` implies `x isa Kd`).

File: base/dict.jl [hunk 1]
```diff
@@ -274,6 +274,60 @@ function empty!(h::Dict{K,V}) where V where K
...
+    dst.maxprobe = src.maxprobe
+    return dst
+end
+
+function copy!(dst::Dict{Kd}, src::Dict{Ks}) where {Ks,Kd>:Ks}
+    try
+        return unsafe_copy!(dst, src)
+    catch
+        empty!(dst)  # avoid leaving `dst` in a corrupt state
+        rethrow()
```
Comment (vtjnash) [inline base/dict.jl:323]:
```suggestion
    finally
        empty!(dst)  # avoid leaving `dst` in a corrupt state
```

File: base/dict.jl [hunk 2]
```diff
@@ -274,6 +274,60 @@ function empty!(h::Dict{K,V}) where V where K
...
+
+    dst.ndel = src.ndel
+    dst.count = src.count
+    dst.age = src.age
+    dst.idxfloor = src.idxfloor
+    dst.maxprobe = src.maxprobe
+    return dst
+end
+
+function copy!(dst::Dict{Kd}, src::Dict{Ks}) where {Ks,Kd>:Ks}
```
Comment (vtjnash) [inline base/dict.jl:318]:
```suggestion
function copy!(dst::Dict{Kd}, src::Dict{Ks}) where {Kd, Ks<:Kd}
```
This order is similar, but slightly more canonical (it clarifies the constraint meaning for matching the sparams)

File: base/dict.jl [hunk 3]
```diff
@@ -274,6 +274,60 @@ function empty!(h::Dict{K,V}) where V where K
...
+    try
+        return unsafe_copy!(dst, src)
+    catch
+        empty!(dst)  # avoid leaving `dst` in a corrupt state
+        rethrow()
+    end
+end
+
+# It's safe to call `unsafe_copy!` if `Vd>:Vs`:
+copy!(dst::Dict{Kd,Vd}, src::Dict{Ks,Vs}) where {Ks,Kd>:Ks,Vs,Vd>:Vs} =
```
Comment (vtjnash) [inline base/dict.jl:328]:
```suggestion
copy!(dst::Dict{Kd,Vd}, src::Dict{Ks,Vs}) where {Kd, Ks<:Kd, Vd, Vs<:Vd} =
```

File: test/dict.jl [hunk 4]
```diff
@@ -1039,13 +1039,35 @@ end
...
-    for a = ([3=>4], [0x3=>0x4], [3=>4, 5=>6, 7=>8], Pair{UInt,UInt}[3=>4, 5=>6, 7=>8])
-        @test s === copy!(s, Dict(a)) == Dict(a)
-        if length(a) == 1 # current limitation of Base.ImmutableDict
-            @test s === copy!(s, Base.ImmutableDict(a[])) == Dict(a[])
+    @testset "copy!(::$(typeof(s)), _)" for s in Any[
+        Dict(1 => 2, 2 => 3),                      # concrete key type
+        Dict{Union{Int,UInt},Int}(1 => 2, 2 => 3), # union key type
+        Dict{Any,Int}(1 => 2, 2 => 3),             # abstract key type
+        Dict{Int,Float64}(1 => 2, 2 => 3),         # values are converted
+    ]
```
Comment (vtjnash) [inline test/dict.jl:1047]: Style nit: indent for `for` and indent for `[` should make at least 2:
```suggestion
            Dict(1 => 2, 2 => 3),                      # concrete key type
            Dict{Union{Int,UInt},Int}(1 => 2, 2 => 3), # union key type
            Dict{Any,Int}(1 => 2, 2 => 3),             # abstract key type
            Dict{Int,Float64}(1 => 2, 2 => 3),         # values are converted
        ]
```

**Review by vtjnash — APPROVED** [review]

File: base/dict.jl (hunk 1 above)
Comment (tkf) [reply base/dict.jl:323]: Doing this in `finally` would mean to empty `dst` always, right?

Comment (vtjnash) [reply base/dict.jl:323]: Oh, right. Carry on.

**Comment by Keno** [comment]
Seems to have broken CI. I'm gonna suggest a revert until this can be investigated.

**Comment by tkf** [comment]
Sounds good to me. The CI was all green as of the last commit so I thought it was good to go.

**Pull request merged**